timelock/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+// This contract demonstrates 'timelock' concept and implements a
+// greatly simplified Claimable Balance (similar to
+// https://developers.stellar.org/docs/glossary/claimable-balance).
+// The contract allows to deposit some amount of token and allow another
+// account(s) claim it before or after provided time point.
+#![no_std]
+#[cfg(feature = "testutils")]
+extern crate std;
+
+use soroban_auth::{
+    check_auth, NonceAuth, {Identifier, Signature},
+};
+use soroban_sdk::{contractimpl, contracttype, BigInt, BytesN, Env, IntoVal, Symbol, Vec};
+use soroban_token_contract as token;
+use token::TokenClient;
+
+#[derive(Clone)]
+#[contracttype]
+pub enum DataKey {
+    Balance,
+    Nonce(Identifier),
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub enum TimeBoundKind {
+    Before,
+    After,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct TimeBound {
+    pub kind: TimeBoundKind,
+    pub timestamp: u64,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct ClaimableBalance {
+    pub token: BytesN<32>,
+    pub amount: BigInt,
+    pub claimants: Vec<Identifier>,
+    pub time_bound: TimeBound,
+}
+
+pub struct ClaimableBalanceContract;
+
+// The 'timelock' part: check that provided timestamp is before/after
+// the current ledger timestamp.
+fn check_time_bound(env: &Env, time_bound: &TimeBound) -> bool {
+    let ledger_timestamp = env.ledger().timestamp();
+
+    match time_bound.kind {
+        TimeBoundKind::Before => ledger_timestamp <= time_bound.timestamp,
+        TimeBoundKind::After => ledger_timestamp >= time_bound.timestamp,
+    }
+}
+
+// Contract usage pattern (pseudocode):
+// 1. Depositor calls `token.approve(depositor_auth, claimable_balance_contract, 100)`
+//    to allow contract to withdraw the needed amount of token.
+// 2. Depositor calls `deposit(depositor_auth, token_id, 100, claimants, time bound)`. Contract
+//    withdraws the provided token amount and stores it until one of the claimants
+//    claims it.
+// 3. Claimant calls `claim(claimant_auth)` and if time/auth conditons are passed
+//    receives the balance.
+#[contractimpl]
+impl ClaimableBalanceContract {
+    pub fn deposit(
+        env: Env,
+        from: Signature,
+        token: BytesN<32>,
+        amount: BigInt,
+        claimants: Vec<Identifier>,
+        time_bound_kind: TimeBoundKind,
+        timestamp: u64,
```

Comment:
💡 These two fields seem to be the same thing as the `TimeBound` type, which suggests you could use that type here.
```suggestion
        time_bound: TimeBound,
```

Reply:
Done.